File: tests/07_phases_hlt.dg/success_collapse_03.cpp
/*
    The aim of this test is to ensure that induction variables are properly
    adjusted after the collapse HLT transform
*/

/*
<testinfo>
test_generator="config/mercurium-hlt run"
</testinfo>
*/

#include <cassert>


int main()
{
    int i, k;
    unsigned int counter = 0;

    #pragma hlt collapse(3)
    for (i = -10; i < 10; i += 6)
    {
        for (int j = 10; j < 0; --j)
        {
            for (k = 10; k < -10; k -= 6)
            {
                counter++;
            }
        }
    }

    assert(i == 14 && k == -14);
}

File: tests/07_phases_hlt.dg/success_collapse_03.f90
!
! Fortran version of test 'success_collapse_03.cpp'
!

! <testinfo>
! test_generator="config/mercurium-hlt run"
! </testinfo>

PROGRAM P
    IMPLICIT NONE
    INTEGER(1) :: I, J, K
    INTEGER(4) :: COUNTER

    COUNTER = 0

    !$HLT COLLAPSE(3)
    DO I = -10, 10, 6
        DO J = 10, 0, -1
            DO K = 10, -10, -6
                COUNTER = COUNTER + 1
            END DO
        END DO
    END DO
    !$HLT END COLLAPSE

    IF (I .NE. 14 .OR. J .NE. -1 .OR. K .NE. -14) THEN
        WRITE(0, *) "Assertion `I .NE. 14 .OR. J .NE. -1 .OR. K .NE. -14' failed."
        STOP
    END IF
END PROGRAM
